cirq-google/cirq_google/ops/z_gate_family.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+
+import cirq
+from cirq.ops import raw_types
+from cirq_google.ops.physical_z_tag import PhysicalZTag
+
+
+_PHYSICAL_Z_TAG = PhysicalZTag()
+
+
+class ZGateFamily(cirq.GateFamily):
```

Comment:
I think instead of creating a new gate family specific to Z gates, we can go and add another keyword parameter to the base `GateFamily` class that takes a list of tags to accept. 

So, the changes to the gate family class would look as follows:

```python
class GateFamily:
    def __init__(
        self,
        gate: Union[Type[raw_types.Gate], raw_types.Gate],
        *,
        name: Optional[str] = None,
        description: Optional[str] = None,
        ignore_global_phase: bool = True,
        tags_to_accept: Optional[Sequence[Hashable]] = None,
    ) -> None:
        self.tags_to_accept = frozenset(tags_to_accept) if tags_to_accept is not None else None
        ...

    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
        if isinstance(item, raw_types.Operation):
            if item.gate is None:
                return False
            if (
                self.tags_to_accept is not None
                and isinstance(item, raw_types.TaggedOperation)
                and self.tags_to_accept.isdisjoint(item.tags)
            ):
                return False
            item = item.gate
        return self._predicate(item)

```


This will give us the convenience to extend support for accepting / ignoring tagged operations for all gate type / instances instead of doing it for a Z gates specifically. Since the default is None, all existing uses of gate family class should be un-affected.

Reply:
This sounds great! My only suggestion is to have both `tags_to_accept` (only accept gates that include this tag) and `tags_to_reject` (reject gates that include this tag). Tags not included in either list don't influence whether the operation is accepted.

So a typical device gateset will instead have `cirq.GateFamily(cirq.ZPowGate, tags_to_accept=[PhysicalZTag()])` to represent the physical Z gate and `cirq.GateFamily(cirq.ZPowGate, tags_to_reject=[PhysicalZTag()])` as the virtual Z gate. If physical Z is off, exclude the physical Z gate family from the gateset.

I'm also working on `PhasedXZGateFamily` which needs to reject ZPowGates with a `PhysicalZTag` if physical Z is off. `super().__init__(..., tags_to_reject=[PhysicalZTag()])` should suffice, and it'll defer to the physical Z gate family to accept PhysicalZTag.

Reply:
If `tags_to_accept` is set, should the GateFamily accept if *all* tags in `tags_to_accept` appear in the `TaggedOperation`, or just *at least one*? In other words, does `tags_to_accept` have AND or OR semantics?

For `PhysicalZTag` this doesn't matter. Are there other use cases for this other than the PhysicalZTag today? If not, I would prefer adding a subclass `TaggedGateFamily` in cirq-google instead, and move to cirq-core only when there are more use cases to inform the design of stable API fields.

Reply:
`tags_to_accept` should have an `OR` semantic -- that's what we've been using in transformers so far as well. 

>  I would prefer adding a subclass TaggedGateFamily in cirq-google instead

IIUC, cirq-google is also supposed to be stable after 1.0 launch, and therefore we won't be able to remove a public-facing API if we end up finding a use case for tags in cirq-core. I also think that adding support for tags in gate families aligns well with the recent effort of adding support for tags in transformers. 

I'd suggest directly making the changes in cirq-core. We can start with adding either of `tags_to_accept` or `tags_to_ignore` if adding only one of the two suffices our current use case. We can also add both if we believe one of them is not enough.

Reply:
> tags_to_accept should have an OR semantic -- that's what we've been using in transformers so far as well.
Ack

> IIUC, cirq-google is also supposed to be stable after 1.0 launch, and therefore we won't be able to remove a public-facing API if we end up finding a use case for tags in cirq-core.
IMO cirq-google should be flexible until QCS is more stable and there were some preferences for this in my discussions with some folks, but I do think this needs to be discussed and communicated more broadly. Will work on that.

But adding directly to core sgtm since the AND vs OR semantic ambiguity is now resolved.

> We can start with adding either of tags_to_accept or tags_to_ignore if adding only one of the two suffices our current use case. We can also add both if we believe one of them is not enough.
If by `tags_to_ignore` you mean the same as what I mean by `tags_to_reject` (i.e. if at least one tag in the collection exists in the TaggedOperation, the operation should be rejected), then we do need it, for reasons in https://github.com/quantumlib/Cirq/pull/5063#discussion_r826415632.

If you mean "tags in the collection don't make a difference in accepting/rejecting a TaggedOperation", then I agree that we don't need it if we already have `tags_to_accept` and `tags_to_ignore`.